### ISSUE TITLE
Possible Data Race due to Memory Ordering in Shared Array

### ISSUE THREAD

**Issue by jtramm** [openmc-dev/openmc#1763]
In our `SharedArray` data structure we make use of a few OpenMP atomic operations. Given that conditionals are involved on the results of these atomics, it's possible that we may have issues with memory ordering. Here's the relevant code:

```C++
  int64_t thread_safe_append(const T& value)
  {
    // Atomically capture the index we want to write to
    int64_t idx;
    #pragma omp atomic capture
    idx = size_++;

    // Check that we haven't written off the end of the array
    if (idx >= capacity_) {
      #pragma omp atomic write
      size_ = capacity_;
      return -1;
    }

    // Copy element value to the array
    data_[idx] = value;

    return idx;
  }
```

By default, OpenMP uses a relaxed memory ordering, which I think in this case could potentially result in undefined behavior. There's some more discussion on memory ordering problems in the [Boehm paper](https://www.usenix.org/legacy/event/hotpar11/tech/final_files/Boehm.pdf).

To guarantee correctness here, I think we just need to add `seq_cst` to our pragmas. The downside is the memory ordering atomic pragmas were only added in OpenMP 4.0, so it's possible use of them may prevent compilation on older compilers. I'm doing some testing now with gcc to see where support drops.

**Comment by jtramm** [comment]
In my testing, gcc 4.8.5 gives an error when using `seq_cst` and fails compilation. gcc 5.5 and above work fine though. It looks like it's possible to query the OpenMP version number from the compiler via the `_OPENMP` macro, so one option might be to use this macro to decide between two sets of pragmas based on what version is available. It would mean older compilers could be subject to the race condition but would at least compile fine.

**Comment by jtramm** [comment]
Closed by #1764.
